### ISSUE TITLE
On connect callback

### ISSUE THREAD

**Issue by tleish** [anycable/anycable-client#17]
I have a scenario where a URL can return a 304 with the prior cached page including an AnyCable JWT token.  In the event of a cached page, when the browser renders the cached page with an expired JWT token, it cannot connect.

From the examples I created the following to update the token:

```javascript
import { createCable } from '@anycable/web';

export default createCable({
  tokenRefresher: async (transport) => {
    let response = await fetch('/cable_url.json');
    let data = await response.json();

    // Update URL for the underlying transport
    transport.setURL(data['url']);
  },
});
```

This works at updating the token when the prior cached token expired.  However, when combined with `<turbo-cable-stream-source />` the channels are not initiated since the attempt to establish the channels occurred prior when the elements were first added to the DOM.  As a workaround, I added the following to remove and re-add the elements.

```diff
export default createCable({
  tokenRefresher: async (transport) => {
    //...

+   // Remove and re-add turbo-cable-stream-source elements to reconnect to channel
+   document.querySelectorAll('turbo-cable-stream-source').forEach((element) => {
+     element.parentNode.replaceChild(element, element);
+   });
  },
});
```

When testing the above code, the elements try to reconnect before the new connection is established and I get an error:

```
index.js:257 Uncaught (in promise) NoConnectionError: No connection
    at Cable._subscribe (index.js:257:46)
    at Cable.subscribe (index.js:245:31)
    at Cable.subscribeTo (index.js:466:17)
    at TurboStreamSourceElement.connectedCallback (stream_source_element.js:12:54)
    at cable.js:17:28
    at NodeList.forEach (<anonymous>)
    at tokenRefresher (cable.js:16:62)
    at async index.js:117:7
    at async index.js:136:11
```

Wrapping it in a `setTimeout` resolves the issue, but feels a hacky.  
```diff
export default createCable({
  tokenRefresher: async (transport) => {
    //...

    // Remove and re-add turbo-cable-stream-source elements to reconnect to channel
+   setTimeout(() => {
      document.querySelectorAll('turbo-cable-stream-source').forEach((element) => {
        element.parentNode.replaceChild(element, element);
      });
+   }, 1);
  },
});
```

Is there a better approach, for example callback available to initiate this when the connection is established?

**Comment by palkan** [comment]
Thanks for the report!

I think, we should make `cable.subscribe(channel)` kind of pessimistic: wait for subscription rejection/confirmation independently of the cable state.

I've started working on this, but that's turned out a bit more trickier than I expected; hope to release a new version later this week.

**Comment by palkan** [comment]
Released 0.4.0 (`core`, `web`, other packages use custom versioning but also updated to match the core dependency).

The most important change is that `cable.subscribe(channel)` automatically handles network issues and keeps the channel subscribed until `channel.disconnect()`  or `cable.disconnect()` (or server sent `disconnect` without `reconnect`) is called.

Please, let me know if there are still some issues.

**Comment by tleish** [comment]
Just tried it out.  I still get the "Uncaught (in promise) DisconnectedError: Disconnected" error in the dev console, but otherwise working, but it appears to be working without the additional logic of refreshing turbo-stream elements.

**Comment by palkan** [comment]
Do you see this error when calling `cable.connect()`?

**Comment by tleish** [comment]
I see this error when the page initially loads with a cached/expired token and the javascript renews the token.  The code does renew the token, but the library also logs this error in the javascript console.

```javascript
// cable.js
import { createCable, fetchTokenFromHTML } from '@anycable/web'

// By default, the current page is loaded in the background,
// and the action-cable-url (or cable-url) meta tag is used to update
// the connection url
export default createCable({tokenRefresher: fetchTokenFromHTML()})
```

![promise-error](https://user-images.githubusercontent.com/1099993/178634491-d89b661f-0761-469d-9f20-0f1b50356396.gif)

**Comment by tleish** [comment]
![image](https://user-images.githubusercontent.com/1099993/178636192-7d445ffe-2759-4309-be62-9b357ad38c88.png)

**Comment by palkan** [comment]
Thanks for the reproduction!

Yeah, that's where the error is initiated. The question is which promise is rejected and not handled. (I'm pretty sure it's `await subscribeTo(...)`, but not yet sure why).

**Comment by tleish** [comment]
The following change removes the error for me, but not sure if it's the right change.

```diff
export class Cable {
  //...
  subscribeTo(ChannelClass, params) {
      //...
-    return this.subscribe(channel).then(() => channel)
+    return this.subscribe(channel).then(() => channel).catch(() => channel)
  }
}
```

**Comment by palkan** [comment]
Just released 0.4.1.

Thanks for you help!
